---
title: "Analysis of GCNT2 Expression Profile"
author: "A. Mihaylov <aleksandar.mihaylov@epfl.ch>"
---

# Preparing Finely Annotated Dataset

## Downloading Finely Annotated Dataset

```{r downloading_processed_dataset}
library(dplyr)
source("../Rutils/functions.R")

ds_name <- "SalasA_2023_GSE214695"
home.path <- file.path(getwd(), "..", ds_name)
output_data_path <- file.path(home.path, "output")

size <- "whole" # "light" or "whole"

# cmc stands for colonic mucosa cells
cmc <- readRDS(file.path(output_data_path, paste0(ds_name, 
                                                  sprintf("_finely_annotated_%s.rds", size))))
```

## Extracting GCNT2 Counts 

We extract the GCNT2 normalized counts and add them as a metadata column in our Seurat object for easier manipulation in downstream analyses.

```{r extracting_counts}
counts <- Seurat::GetAssayData(cmc, assay = "RNA", slot = "data")
gcnt2_counts <- counts["GCNT2", ]
cmc$GCNT2 <- gcnt2_counts
```

# Initial Analysis

## Feature Plot

We begin by analyzing whether GCNT2 expression is restricted to certain regions within the Umap plot.

```{r in_feature_plot}
# Feature plot of GCNT2 on Umap reduction
Seurat::FeaturePlot(cmc, features = "GCNT2") + 
  ggplot2::ggtitle("Feature Plot of GCNT2 Expression across Cell Types")
```

Seemingly, GCNT2 is expressed by very few cells in our Seurat object. Still, GCNT2 expression seems quite abundant in a sub-cluster of B cells. Further investigations on that matter can be found in the file *gcnt2_bcell_expression.Rmd*. 

For now, let us quantify the number and percentage of cells, within our Seurat object, that express GCNT2.

## Number of Cells Expressing GCNT2

We begin by creating a new binary metadata column indicating whether a cell expresses GCNT2.

```{r binary_gcnt2}
cmc$binary_gcnt2 <- ifelse(cmc$GCNT2 > 0, TRUE, FALSE)
```

```{r gcnt2_expressing_cells}
cat("Number of cells in Seurat object:", length(Seurat::Cells(cmc)), "\n")
cat("Number of cells, within Seurat object, that express GCNT2:", sum(cmc$binary_gcnt2),
    paste0("(", round(sum(cmc$binary_gcnt2)*100/length(Seurat::Cells(cmc)), 3), "%)"))
```

Clearly, most cells, do not express GCNT2. Nevertheless, is this behavior independent of cell type, gut condition, or individual. Could it be that certain cell types, gut conditions, or individuals express GCNT2 more than others?

# Binary GCNT2 Expression Analysis

To answer the above question, consider that each cell either expresses GCNT2, at some non-zero level, or doesn't. This enables us to classify cells in a binary fashion: TRUE if GCNT2 is expressed and FALSE if it isn't. Subsequently, we can perform statistical tests to check for significance between cell types, gut conditions, and individuals.

## Per Cell Type 

```{r cell_type_binary_expression}
dittoSeq::dittoBarPlot(
    object = cmc,
    var = "binary_gcnt2",
    group.by = "manual_annotation") + 
  ggplot2::scale_fill_manual(values = c("TRUE" = "green", "FALSE" = "indianred2"),
                             name = "GCNT2 Expression") +
  ggplot2::ggtitle("Stacked Bar Plot of Percent of Cells Expressing GCNT2 per Cell Type") +
  ggplot2::xlab("Cell Type") +
  ggplot2::ylab("Percent of Cells")

cmc@meta.data %>% 
  dplyr::group_by(manual_annotation) %>% 
  summarise(total_cells = n(),
            gcnt2_expressing_cells = sum(binary_gcnt2),
            expression_ratio = gcnt2_expressing_cells*100/total_cells) -> cell_type_binary_expression
  
cell_type_binary_expression
```

## Per Gut Condition

```{r condtion_binary_expression}
dittoSeq::dittoBarPlot(
    object = cmc,
    var = "binary_gcnt2",
    group.by = "Treatment") + 
  ggplot2::scale_fill_manual(values = c("TRUE" = "green", "FALSE" = "indianred2"),
                             name = "GCNT2 Expression") +
  ggplot2::ggtitle("Stacked Bar Plot of Percent of Cells Expressing GCNT2 per Gut Condition") +
  ggplot2::xlab("Gut Condition") +
  ggplot2::ylab("Percent of Cells")

cmc@meta.data %>% 
  dplyr::group_by(Treatment) %>% 
  summarise(total_cells = n(),
            gcnt2_expressing_cells = sum(binary_gcnt2),
            expression_ratio = gcnt2_expressing_cells*100/total_cells) -> treatment_binary_expression
  
 treatment_binary_expression
```

## Per Cell Type & Gut Condition

```{r cell_type_condition_binary_expression}
cmc@meta.data %>% 
  dplyr::group_by(manual_annotation, Treatment) %>% 
  summarise(
    total_cells = n(),
    gcnt2_expressing_cells = sum(binary_gcnt2),
    expression_ratio = gcnt2_expressing_cells*100/total_cells
  ) -> cell_type_treatment_binary_expression

ggplot2::ggplot(cell_type_treatment_binary_expression, ggplot2::aes(x = Treatment, y = manual_annotation,
                                                                     fill = expression_ratio)) +
  ggplot2::geom_tile() +
  ggplot2::geom_text(ggplot2::aes(label = round(expression_ratio, 3)), color = "black", size = 4) +
  ggplot2::scale_fill_gradient(low = "yellow", high = "darkolivegreen3",
                               name = "Expression Ratio") +
  ggplot2::ggtitle("Heatmap of GCNT2 Expression Ratio across 
                   Cell Types and Gut Conditions") +
  ggplot2::xlab("Gut Condition") +
  ggplot2::ylab("Cell Type") +
  ggplot2::theme_classic()
```

## Per Cell Type & Individual

```{r cell_type_individual_binary_expression}
cmc@meta.data %>% 
  dplyr::group_by(manual_annotation, Sample) %>% 
  summarise(
    total_cells = n(),
    gcnt2_expressing_cells = sum(binary_gcnt2),
    expression_ratio = gcnt2_expressing_cells*100/total_cells
  ) -> cell_type_sample_binary_expression

ggplot2::ggplot(cell_type_sample_binary_expression, ggplot2::aes(x = manual_annotation, y = Sample,
                                                     fill = expression_ratio)) +
  ggplot2::geom_tile() +
  ggplot2::geom_text(ggplot2::aes(label = round(expression_ratio, 3)), color = "black", size = 3) +
  ggplot2::scale_fill_gradient(low = "yellow", high = "darkolivegreen3",
                               name = "Expression Ratio") +
  ggplot2::ggtitle("Heatmap of GCNT2 Expression Ratio across 
                   Cell Types and Individuals") +
  ggplot2::xlab("Cell Type") +
  ggplot2::ylab("Individuals") + 
  ggplot2::theme_classic()
```

Having the expression ratio data frame across individuals and cell types, as constructed above, we can perform statistical tests by grouping the individuals per gut condition. This way, we will be able to infer whether GCNT2 expression changes per cell type, across conditions. 

```{r testing_expression_ratio}
cell_type_sample_binary_expression$Condition <- substr(cell_type_sample_binary_expression$Sample, 1, 2)
cell_types = unique(cell_type_sample_binary_expression$manual_annotation)

# Anova tables per cell type across conditions
for(cell_type in cell_types) {
  cells_of_interest <- filter(cell_type_sample_binary_expression, manual_annotation == cell_type)
             
  condition_frequency <- table(cells_of_interest$Condition) 
  
  un_rep_conditions <- names(condition_frequency)[condition_frequency <= 1]
  
  if(length(un_rep_conditions) < length(unique(cells_of_interest$Condition))) {
    cells_of_interest <- filter(cells_of_interest, !(Condition %in% un_rep_conditions))
    gcnt2.aov.model <- aov(cells_of_interest$expression_ratio ~ cells_of_interest$Condition)
    cat(paste0("\033[31m", cell_type, "\033[39m\n"))
    print(summary(gcnt2.aov.model))
    cat("\n")
  }
}

# Significance plots per cell type across conditions
for(cell_type in cell_types){
  cells_of_interest <- filter(cell_type_sample_binary_expression, manual_annotation == cell_type)
  print(signif.plot(cells_of_interest, x = "Condition", y = "expression_ratio", plot_type = "Scatter-Boxplot",
              test = "Wilcox", plot_title = cell_type))
}
```

# Analysis per Cell Type 

## Volcano Plot

The Volcano Plot will enable us to visually situate GCNT2 on the ladder of differential gene expression, grouped by cell type.

```{r de_cell_type}
cmc.markers.cell_type <- Seurat::FindAllMarkers(cmc, test.use = "wilcox", group.by = "manual_annotation",
                                                max.cells.per.ident = Inf, random.seed = 42)

cmc.markers.cell_type$de_expression <- "NO"
cmc.markers.cell_type$de_expression[cmc.markers.cell_type$avg_log2FC > 0.6 
                                    & cmc.markers.cell_type$p_val_adj < 0.05] <- "UP"
cmc.markers.cell_type$de_expression[cmc.markers.cell_type$avg_log2FC < -0.6 
                                    & cmc.markers.cell_type$p_val_adj < 0.05] <- "DOWN"

cmc.markers.cell_type$GCNT2_label <- NA
cmc.markers.cell_type["GCNT2", "GCNT2_label"] <- "GCNT2"
```

```{r volcano_plot_cell_type}
color_palette <- c("NO" = "black", "UP" = "green", "DOWN" = "red")

ggplot2::ggplot(cmc.markers.cell_type, ggplot2::aes(x = avg_log2FC, y = -log10(p_val_adj), 
                                                    col = de_expression, label = GCNT2_label)) +
  ggplot2::geom_point(size = 0.3) +
  ggplot2::scale_color_manual(values = color_palette) +
  ggplot2::labs(color = "Differential Expression") +
  ggrepel::geom_label_repel() +
  ggplot2::geom_vline(xintercept=c(-0.6, 0.6), col="magenta") +
  ggplot2::geom_hline(yintercept=-log10(0.05), col="magenta") + 
  ggplot2::ggtitle("Volcano Plot, Grouped by Cell Type") +
  ggplot2::theme_classic()
```

GCNT2 is 'DOWN' differentially expressed. We can retrieve its p-value and avg_log2FC.

```{r gcnt2_stats_cell_type}
cat("Differential Expression of GCNT2:", cmc.markers.cell_type["GCNT2", "de_expression"], "\n")
cat("GCNT2 p-value:",  cmc.markers.cell_type["GCNT2", "p_val_adj"], "\n")
cat("GCNT2 avg_log2FC:",  cmc.markers.cell_type["GCNT2", "avg_log2FC"], "\n")
```

## Statistical Tests

```{r stat_tests_per_cell_type}
# Anova table across cell types
gcnt2.aov.model <- aov(cmc@meta.data$GCNT2 ~ cmc@meta.data$manual_annotation)
print(as.data.frame(summary(gcnt2.aov.model)[[1]]))

# Post-hoc Tukey HSD test for all pairs of cell types
gcnt2.tukey.results <- TukeyHSD(gcnt2.aov.model)
print(as.data.frame(gcnt2.tukey.results[[1]]))

# Wilcoxon test for all pairs of cell types
gcnt2.wilcox.results <- multiple.wilcox.tests(cmc@meta.data, 
                                              "manual_annotation", 
                                              "GCNT2") 
print(gcnt2.wilcox.results)

# Two-sample t-test for all pairs of cell types
gcnt2.t_test.results <- pairwise.t.test(cmc@meta.data$GCNT2,
                                        cmc@meta.data$manual_annotation, 
                                        alternative = "two.sided", 
                                        p.adjust.method = "fdr")
print(as.data.frame(gcnt2.t_test.results$p.value))
```

## Diagnostic Plots

```{r plots_per_cell_type}
# Violin plots of GCNT2 expression across cell types
Seurat::VlnPlot(cmc, "GCNT2", group.by = "manual_annotation") +
  ggplot2::ggtitle("Violin Plots of GCNT2 Expression across Cell Types") 

# Box plots of GCNT2 expression across cell types
ggplot2::ggplot(cmc@meta.data, 
                ggplot2::aes(x = manual_annotation, y = GCNT2)) + 
  ggplot2::geom_boxplot(fill="slateblue", alpha=0.2) + 
  ggplot2::xlab("Cell Type") +
  ggplot2::ylab("GCNT2 Expression") +
  ggplot2::ggtitle("Boxplot of GCNT2 Expression across Cell Types")

# Ridge plots of GCNT2 expression across cell types
Seurat::RidgePlot(cmc, features = "GCNT2", group.by = "manual_annotation") +
  ggplot2::ggtitle("Ridge Plot of GCNT2 Expression across Cell Types")
```

## Significance Plots

In this sub-section, we combine both the diagnostic plots and the test results that were obtained into significance plots. In these plots, one can visualize GCNT2 expression per cell type, and observe the statistical difference in GCNT2 expression between cell types. 

*Significance levels*

| P-value  | Legend  |
|----------|---------|
| > 0.05   | Nothing |
| ≤ 0.05   | *       |
| ≤ 0.01   | **      |
| ≤ 0.001  | ***     |
| ≤ 0.0001 | ****    |



```{r significance_plots_per_cell_type}
signif.plot(cmc@meta.data, x = "manual_annotation", y = "GCNT2", 
            plot_type = "Violin", test = "Wilcox",
            plot_title = "GCNT2 Expression vs. Cell Type, Wilcoxon Test")

signif.plot(cmc@meta.data, x = "manual_annotation", y = "GCNT2", 
            plot_type = "Boxplot", test = "T-test",
            plot_title = "GCNT2 Expression vs. Cell Type, T-Test")
```

# Analysis per Condition

## Volcano Plot

The Volcano Plot will enable us to visually situate GCNT2 on the ladder of differential gene expression, grouped by gut condition.

```{r de_condition}
cmc.markers.treatment <- Seurat::FindAllMarkers(cmc, test.use = "wilcox", group.by = "Treatment",
                                      max.cells.per.ident = Inf, random.seed = 42)

cmc.markers.treatment$de_expression <- "NO"
cmc.markers.treatment$de_expression[cmc.markers.treatment$avg_log2FC > 0.6 
                                    & cmc.markers.treatment$p_val_adj < 0.05] <- "UP"
cmc.markers.treatment$de_expression[cmc.markers.treatment$avg_log2FC < -0.6 
                                    & cmc.markers.treatment$p_val_adj < 0.05] <- "DOWN"

cmc.markers.treatment$GCNT2_label <- NA
cmc.markers.treatment["GCNT2", "GCNT2_label"] <- "GCNT2"
```

```{r volcano_plot_condition}
color_palette <- c("NO" = "black", "UP" = "green", "DOWN" = "red")

ggplot2::ggplot(cmc.markers.treatment, ggplot2::aes(x = avg_log2FC, y = -log10(p_val_adj), 
                                                    col = de_expression, label = GCNT2_label)) +
  ggplot2::geom_point(size = 0.3) +
  ggplot2::scale_color_manual(values = color_palette, na.translate = FALSE) +
  ggplot2::labs(color = "Differential Expression") +
  ggrepel::geom_label_repel() +
  ggplot2::geom_vline(xintercept=c(-0.6, 0.6), col="magenta") +
  ggplot2::geom_hline(yintercept=-log10(0.05), col="magenta") + 
  ggplot2::ggtitle("Volcano Plot, Grouped by Gut Condition") +
  ggplot2::theme_classic()
```

In the above Volcano Plot, GCNT2 does not show up because its p-value is NA.

```{r gcnt2_stats_treatment}
cat("Differential Expression of GCNT2:", cmc.markers.treatment["GCNT2", "de_expression"], "\n")
cat("GCNT2 p-value:",  cmc.markers.treatment["GCNT2", "p_val_adj"], "\n")
cat("GCNT2 avg_log2FC:",  cmc.markers.treatment["GCNT2", "avg_log2FC"], "\n")
```

## Statistical Tests

```{r stat_tests_per_condition}
# Anova table across gut conditions
gcnt2.aov.model <- aov(cmc@meta.data$GCNT2 ~ cmc@meta.data$Treatment)
print(as.data.frame(summary(gcnt2.aov.model)[[1]]))

# Post-hoc Tukey HSD test for all pairs of gut conditions
gcnt2.tukey.results <- TukeyHSD(gcnt2.aov.model)
print(as.data.frame(gcnt2.tukey.results[[1]]))

# Wilcoxon test for all pairs of gut conditions
gcnt2.wilcox.results <- multiple.wilcox.tests(cmc@meta.data, 
                                              "Treatment", 
                                              "GCNT2") 
print(gcnt2.wilcox.results)

# Two-sample t-test for all pairs of gut conditions
gcnt2.t_test.results <- pairwise.t.test(cmc@meta.data$GCNT2,
                                        cmc@meta.data$Treatment, 
                                        alternative = "two.sided", 
                                        p.adjust.method = "fdr")
print(as.data.frame(gcnt2.t_test.results$p.value))
```

## Diagnostic Plots

```{r plots_per_condition}
# Violin plots of GCNT2 expression across gut conditions
Seurat::VlnPlot(cmc, "GCNT2", group.by = "Treatment") +
  ggplot2::ggtitle("Violin Plots of GCNT2 Expression across Conditions") 

# Box plots plots of GCNT2 expression across gut conditions
ggplot2::ggplot(cmc@meta.data, 
                ggplot2::aes(x = Treatment, y = GCNT2)) + 
  ggplot2::geom_boxplot(fill="slateblue", alpha=0.2) + 
  ggplot2::xlab("Condition") +
  ggplot2::ylab("GCNT2 Expression") +
  ggplot2::ggtitle("Boxplot of GCNT2 Expression across Conditions")

# Ridge plots of GCNT2 expression across gut conditions
Seurat::RidgePlot(cmc, features = "GCNT2", group.by = "Treatment") +
  ggplot2::ggtitle("Ridge Plot of GCNT2 Expression across Conditions")
```

## Significance Plots

In this sub-section, we combine both the diagnostic plots and the test results that were obtained into significance plots. In these plots, one can visualize GCNT2 expression per gut condition, and observe the statistical difference in GCNT2 expression between gut conditions. 

*Significance levels*

| P-value  | Legend  |
|----------|---------|
| > 0.05   | Nothing |
| ≤ 0.05   | *       |
| ≤ 0.01   | **      |
| ≤ 0.001  | ***     |
| ≤ 0.0001 | ****    |

```{r significance_plots_per_condition}
signif.plot(cmc@meta.data, x = "Treatment", y = "GCNT2", 
            plot_type = "Scatter", test = "Wilcox",
            plot_title = "GCNT2 Expression vs. Gut Condition, Wilcoxon Test")

signif.plot(cmc@meta.data, x = "Treatment", y = "GCNT2", 
            plot_type = "Boxplot", test = "Tukey",
            plot_title = "GCNT2 Expression vs. Gut Condition, Tukey HSD Test")
```

# Analysis per Cell Type per Gut Condition

In this section, we iterate over each cell type and subdivide it into the 3 distinct gut conditions. This way, we will analyze GCNT2 expression per cell type per gut condition.

```{r setting_idents}
Seurat::Idents(cmc) <- "manual_annotation"

cell_types <- unique(Seurat::Idents(cmc))
conditions <- unique(cmc$Treatment)
```

## Heatmap

```{r dot_plot}
cmc@meta.data %>% 
  dplyr::group_by(manual_annotation, Treatment) %>% 
  summarise(mean_GCNT2 = mean(GCNT2), .groups = "drop") -> mean.gcnt2.df 

ggplot2::ggplot(mean.gcnt2.df, ggplot2::aes(x = Treatment, y = manual_annotation, fill = mean_GCNT2)) +
  ggplot2::geom_tile() +
  ggplot2::scale_fill_gradient(low = "yellow", high = "red") +
  ggplot2::labs(title = "Heatmap of GCNT2 Expression across Cell Types and Gut Conditions",
                fill = "Mean GCNT2 expression") +
  ggplot2::xlab("Gut Condition") +
  ggplot2::ylab("Cell Type") +
  ggplot2::theme_classic()
```

Again, we see that B cells seem to upregulate GCNT2 expression under certain conditions. For further analysis, check out the *gcnt2_bcell_expression.Rmd* file.

## Anova Tables

```{r anova_tables}
for(cell_type in cell_types){
  cell_type_meta <- subset(cmc, idents = cell_type)@meta.data
  condition_frequency <- table(cell_type_meta$Treatment) 
  
  un_rep_conditions <- names(condition_frequency)[condition_frequency <= 1]
  
  if(length(un_rep_conditions) < length(unique(cell_type_meta$Treatment))) {
    cell_type_meta <- subset(cell_type_meta, subset = !(Treatment %in% un_rep_conditions))
    gcnt2.aov.model <- aov(cell_type_meta$GCNT2 ~ cell_type_meta$Treatment)
    cat(paste0("\033[31m", cell_type, "\033[39m\n"))
    print(summary(gcnt2.aov.model))
    cat("\n")
  }
}
```

## Wilcoxon Tests

```{r wilcoxon_tests}
for(cell_type in cell_types){
  cell_type_meta <- subset(cmc, idents = cell_type)@meta.data
  condition_frequency <- table(cell_type_meta$Treatment) 
  
  un_rep_conditions <- names(condition_frequency)[condition_frequency <= 1]
  
  if(length(un_rep_conditions) < length(unique(cell_type_meta$Treatment))) {
    cell_type_meta <- subset(cell_type_meta, subset = !(Treatment %in% un_rep_conditions))
    gcnt2.wilcox.results <- multiple.wilcox.tests(cell_type_meta, "Treatment", "GCNT2")
    cat(paste0("\033[31m", cell_type, "\033[39m\n"))
    print(gcnt2.wilcox.results)
    cat("\n")
  }
}
```

## Ridge Plots

```{r ridge_plots}
for(cell_type in cell_types){
  
  cell_type_seurat <- subset(cmc, idents = cell_type)
  condition_frequency <- table(cell_type_seurat$Treatment) 
  
  un_rep_conditions <- names(condition_frequency)[condition_frequency <= 1]
  
  if (length(un_rep_conditions) < length(unique(cell_type_seurat$Treatment))){
    cell_type_seurat <- subset(cell_type_seurat, subset = !(Treatment %in% un_rep_conditions)) 
    
    print(Seurat::RidgePlot(cell_type_seurat, features = "GCNT2", group.by = "Treatment") + 
        ggplot2::ggtitle(cell_type))
  }
}
```

## Significance Plots

In this sub-section, we combine both the diagnostic plots and the test results that were obtained into significance plots. There is a significance plot per cell type, where one can visualize differences in GCNT2 expression across the different gut conditions,and observe their statistical significance.

*Significance levels*

| P-value  | Legend  |
|----------|---------|
| > 0.05   | Nothing |
| ≤ 0.05   | *       |
| ≤ 0.01   | **      |
| ≤ 0.001  | ***     |
| ≤ 0.0001 | ****    |

```{r significance_plots}
for(cell_type in cell_types){
  cell_type_meta <- subset(cmc, idents = cell_type)@meta.data
  condition_frequency <- table(cell_type_meta$Treatment) 
  
  un_rep_conditions <- names(condition_frequency)[condition_frequency <= 1]
  
  if(length(un_rep_conditions) < length(unique(cell_type_meta$Treatment))) {
    cell_type_meta <- subset(cell_type_meta, subset = !(Treatment %in% un_rep_conditions))
    print(signif.plot(cell_type_meta, x = "Treatment", y = "GCNT2", plot_type = "Boxplot",
                      test = "Wilcox", plot_title = cell_type))
  }
}
```

# Pseudobulk 

## Pseudobulk Preparation

Firstly, we aim to aggregate GCNT2 expression per individual and cell type. Thus, for each cell type, we will have 18 values corresponding to the 18 distinct patients.

```{r preparation}
cmc_pseudo <- Seurat::AggregateExpression(cmc, features = "GCNT2", slot = "data",
                                          group.by = c("manual_annotation", "Sample"))
cmc_pseudo <- cmc_pseudo[[1]] %>%
  as.data.frame() %>%
  tidyr::pivot_longer(
    cols = tidyr::everything(),
    names_to = "cell_sample",
    values_to = "GCNT2") %>%
  separate(cell_sample, into = c("cell.type", "sample"), sep = "_")

cmc_pseudo$treatment <- substr(cmc_pseudo$sample, 1, 2)

cell_types <- unique(cmc_pseudo$cell.type)
```

## Wilcoxon Tests

```{r wilcoxon_tests_}
for(cell_type in cell_types){
  cat(cell_type, "\n")
  pseudo_cell_type <- subset(cmc_pseudo, subset = (cell.type == cell_type))
  print(multiple.wilcox.tests(pseudo_cell_type, "treatment", "GCNT2"))
}
```

## Significance Plots per Cell Type per Gut Condition

Let us display again significance scatter plots per cell type and gut condition. The scattered points represent pseudobulk results, per individual, instead of single cells. 

*Significance levels*

| P-value  | Legend  |
|----------|---------|
| > 0.05   | Nothing |
| ≤ 0.05   | *       |
| ≤ 0.01   | **      |
| ≤ 0.001  | ***     |
| ≤ 0.0001 | ****    |

```{r pseudo_significance_plots}
for(cell_type in cell_types){
  pseudo_cell_type <- subset(cmc_pseudo, subset = (cell.type == cell_type))
  print(signif.plot(pseudo_cell_type, x = "treatment", y = "GCNT2", plot_type = "Scatter-Boxplot",
                      test = "Wilcox", plot_title = cell_type))
}
```

# GCNT2 Correlation with Other Genes

## Granzymes 

Let us first investigate the correlation of GCNT2 with the Granzyme genes: Granzyme A, B, H, and K.

```{r granzyme_correlation}
granzymes = c("GZMA", "GZMB", "GZMH", "GZMK")

Seurat::DotPlot(cmc, features = c("GCNT2", granzymes)) + Seurat::RotatedAxis()

for(granzyme in granzymes) {
  print(Seurat::FeatureScatter(cmc, feature1 = "GCNT2", feature2 = granzyme))
}
```

## Other Genes

```{r correlation_with_other_genes}
Seurat::Idents(cmc) <- "manual_annotation"
cell_types <- unique(Seurat::Idents(cmc))

for(cell_type in cell_types){
  cells_of_interest <- subset(cmc, idents = cell_type)
  
  expression_matrix <- as.matrix(Seurat::GetAssayData(cells_of_interest, assay = "RNA", slot = "data"))
  gcnt2_expression <- expression_matrix["GCNT2",]
  
  gcnt2_correlation <- apply(expression_matrix, 1, function(x) {
    suppressWarnings(cor(x, gcnt2_expression, method = "pearson"))
  })
  
  gcnt2_correlation <- gcnt2_correlation[names(gcnt2_correlation) != "GCNT2"]
  
  cat(paste0("\033[31m", cell_type, "\033[39m\n"))
  cat(cat("\033[32mPositively Correlated Genes\033[39m\n"))
  print(head(sort(gcnt2_correlation, decreasing = TRUE), 10))
  cat(cat("\033[32mNegatively Correlated Genes\033[39m\n"))
  print(head(sort(gcnt2_correlation, decreasing = FALSE), 10))
  cat("\n")
}
```

## Deeper Dive: Mast Cells

```{r mast_cells_correlation_plots}
positively_correlated_genes <- c('LAMA2', 'DYNLRB2', 'GP1BA', 'ABCG8', 'CTIF', 'GABRD')

mast_cells <- subset(cmc, idents = "Mast")

Seurat::DotPlot(mast_cells, features = c("GCNT2", positively_correlated_genes)) + Seurat::RotatedAxis()

for(gene in positively_correlated_genes) {
  print(Seurat::FeatureScatter(mast_cells, feature1 = "GCNT2", feature2 = gene, jitter = FALSE))
}
```

## Deeper Dive: Endothelial Cells

```{r endothelial_cells_correlation_plots}
positively_correlated_genes <- c('ZDHHC11B', 'PRDM16-DT', 'MYO7B', 'CGNL1', 'L1CAM', 'HR', 'LINC01504', 'FGF18',                                    'PHF19', 'WDR86') 

endothelial_cells <- subset(cmc, idents = "Endothelial")

Seurat::DotPlot(endothelial_cells, features = c("GCNT2", positively_correlated_genes)) + Seurat::RotatedAxis()

for(gene in positively_correlated_genes) {
  print(Seurat::FeatureScatter(endothelial_cells, feature1 = "GCNT2", feature2 = gene, jitter = FALSE))
}
```

## Filtering Out GCNT2 Non-expressing Cells

```{r filtering_seurat_object}
filtered_cmc <- subset(cmc, subset = GCNT2 > 0)
nrow(filtered_cmc@meta.data)
```

```{r filtered_correlation_with_other_genes}
Seurat::Idents(filtered_cmc) <- "manual_annotation"
cell_types <- unique(Seurat::Idents(filtered_cmc))
gcnt2_correlation <- list()

for(cell_type in cell_types){
  cells_of_interest <- subset(filtered_cmc, idents = cell_type)
  cell_correlation <- data.frame()
  
  if(nrow(cells_of_interest@meta.data) > 1) {
    
    expression_matrix <- as.matrix(Seurat::GetAssayData(cells_of_interest, assay = "RNA", slot = "data"))
    gcnt2_expression <- expression_matrix["GCNT2",]
    
    cor_results <- t(apply(expression_matrix, 1, function(x) {
      gene_corr <- suppressWarnings(cor.test(x, gcnt2_expression, method = "pearson"))
      c(estimate = gene_corr$estimate[["cor"]],
        p_value  = gene_corr$p.value)
    }))
    
    cor_results <- as.data.frame(cor_results)
    cor_results$gene <- rownames(expression_matrix)
    cor_results$cell_type <- cell_type 
    
    cell_correlation <- cell_correlation %>% 
      rbind(cor_results) %>% 
      dplyr::filter(gene != "GCNT2") %>% 
      dplyr::filter(p_value <= 0.05 & (estimate <= -0.6 | estimate >= 0.6)) %>% 
      dplyr::arrange(desc(estimate))
    
    gcnt2_correlation[[cell_type]] <- cell_correlation
    
    print(cell_correlation)
  }
}
```

### Deeper Dive: Intersection

```{r epithelial_cells_correaltion_plots}
dplyr::intersect(gcnt2_correlation[["PlasmaCell"]]$gene, gcnt2_correlation[["Endothelial"]]$gene)
```

```{r cells_correaltion_plots}
correlated_genes <- c("BIRC6", "SMCR8", "MUC20-OT1", "SMC6", "MARK3", "H2BC8", "UBE2D3", "ITGA6",    
"MT-ND5", "NSD1", "MACF1", "RAPGEF1", "HNRNPU", "LUC7L3", "TRMT112", "RPN2") 

plasma_endothelial <- subset(filtered_cmc, idents = c("PlasmaCell", "Endothelial"))

Seurat::DotPlot(filtered_cmc, features = c("GCNT2", correlated_genes)) + Seurat::RotatedAxis()

for(gene in correlated_genes) {
  print(Seurat::FeatureScatter(plasma_endothelial, feature1 = "GCNT2", feature2 = gene, jitter = TRUE))
}
```






